src/storage/mvcc/reader/scanner.rs

```diff
@@ -97,34 +99,408 @@ impl<S: Snapshot> BackwardScannerBuilder<S> {
         self
     }
 
-    /// Build `BackwardScanner` from the current configuration.
-    pub fn build(self) -> Result<BackwardScanner<S>> {
-        let lock_cursor = CursorBuilder::new(&self.snapshot, CF_LOCK)
+    /// Build `ForwardScanner` from the current configuration.
+    pub fn build(self) -> Result<Scanner<S>> {
+        let lock_cursor_builder = CursorBuilder::new(&self.snapshot, CF_LOCK)
             .range(self.lower_bound.clone(), self.upper_bound.clone())
-            .fill_cache(self.fill_cache)
-            .scan_mode(ScanMode::Backward)
-            .build()?;
+            .fill_cache(self.fill_cache);
 
-        let write_cursor = CursorBuilder::new(&self.snapshot, CF_WRITE)
+        let write_cursor_builder = CursorBuilder::new(&self.snapshot, CF_WRITE)
             .range(self.lower_bound.clone(), self.upper_bound.clone())
-            .fill_cache(self.fill_cache)
-            .scan_mode(ScanMode::Backward)
-            .build()?;
+            .fill_cache(self.fill_cache);
+        if self.desc {
+            let lock_cursor = lock_cursor_builder.scan_mode(ScanMode::Backward).build()?;
+            let write_cursor = write_cursor_builder.scan_mode(ScanMode::Backward).build()?;
+            Ok(Scanner::Backward(BackwardScanner::new(
+                self,
+                lock_cursor,
+                write_cursor,
+            )))
+        } else {
+            let write_cursor = write_cursor_builder.build()?;
+            let lock_cursor = lock_cursor_builder.build()?;
+            Ok(Scanner::Forward(ForwardScanner::new(
+                self,
+                lock_cursor,
+                write_cursor,
+            )))
+        }
+    }
+}
+
+/// This struct can be used to scan keys starting from the given user key (greater than or equal).
+///
+/// Internally, for each key, rollbacks are ignored and smaller version will be tried. If the
+/// isolation level is SI, locks will be checked first.
+///
+/// Use `ScannerBuilder` to build `ForwardScanner`.
+pub struct ForwardScanner<S: Snapshot> {
```

Comment:
I don't think it's a good idea to put them together in one file. This file is too large.